object-model/copy-semantics.md

```diff
@@ -0,0 +1,151 @@
+Explains when and how objects are copied.
+
+## Foundational knowledge: How and when are copies made
+
+### Background/Required Knowledge
+
+A student:
+* knows what a type is?             [[C++ object model: types]][1]
+* knows what an object is?          [[C++ object model: objects]][2], [[C++ object model: constant objects]][3]
+* knows what class invariants are?
+
+It helps when a student:
+* knows move semantics              [[C++ object model: move semantics]][4]
+* knows special member functions    [[C++ object model: special member functions]][5]
+
+### Student outcomes
+
+_A list of things “a student should be able to” after the curriculum._
+_The next word should be an action word and testable in an exam._
+_Max 5 items._
+
+A student should be able to:
+
+1. explain what copy semantics accomplish
+  * establishing “equivalent” object state in another object
+2. explain difference between deep and shallow copy
+3. explain where copies are made
+
+### Caveats
+
+_This section mentions subtle points to understand, like anything resulting in
+implementation-defined, unspecified, or undefined behavior._
+
+* Compiler-provided (shallow) copies may result in ownership problems (e.g. `char*`).
+
+### Points to cover
+
+_This section lists important details for each point._
+
+* Principle of copying
+  * Copying of trivially copyable types
+  * Copying of non-trivial class types
+  * Copying an object does not change the original
+* Copy Flavors
+  * Shallow copy
+  * No copy
+  * Deep copy
+* Practical applications
+  * Unique_ptr  (no copy)
+  * Strings (deep copy) 
+
+
+## Advanced: Implementing user-defined copy operations
+
+### Background/Required Knowledge
+
+A student:
+* knows special member functions  [C++ object model: special member functions
+
+It helps when a student:
+* knows move semantics            [[C++ object model: move semantics]][4]
+* knows the rule of five          [[C++ object model: rule-of-five]][6]
+* knows the rule of zero          [[C++ object model: rule-of-zero]][7]
+
+### Student outcomes
+
+_A list of things “a student should be able to” after the curriculum._
+_The next word should be an action word and testable in an exam._
+_Max 5 items._
+
+A student should be able to:
+* explain when they have to implement the copy operations for their own type
+  * Copy constructor
+  * Copy assignment operator
+* implement copy operations for their own types
+* _Optional_: explain when copying with basic and strong exception guarantees is useful.
+
+### Caveats
+
+_This section mentions subtle points to understand, like anything resulting in
+implementation-defined, unspecified, or undefined behavior._
+
+* Intricacies when implementing copy operations:
+  * Maybe: Copying going wrong? (std::auto_ptr …)
+
+### Points to cover
+
+_This section lists important details for each point._
+
+* Copy constructors and copy assignment operators
+  * How compiler generates default copy operations
+  * =default, =delete (No copy)
```

Comment:
Why are `=default` and `=delete` considered "advanced"?

Reply:
I can see how `=delete` would be "advanced," because it's useful only for types that are not copyable (i.e. move-only or immobile), which is a more advanced topic than simply "make everything copyable."

For `=default`, I don't know. You _can_ get by without talking about explicitly defaulted special members (I mean, C++03 did, for decades); but _should_ you?